quint/src/parsing/ToIrListener.ts

```diff
@@ -183,19 +194,22 @@ export class ToIrListener implements QuintListener {
 
   // special case for: nondet x = e1; e2
   exitNondet(ctx: p.NondetContext) {
-    const def = this.declarationStack.pop()
-    const nested = this.exprStack.pop()
+    const def = this.declarationStack.pop() ?? this.undefinedDeclaration(ctx)()
+    const nested = this.exprStack.pop() ?? this.undefinedExpr(ctx)()
+
+    if (def.kind !== 'def') {
+      // only `QuintDef` is allowed in `nondet` expressions
+      return
+    }
```

Comment:
This is already ensured by the parser, so we must already have gathered an error for this that will be reported later.

Reply:
Sorry if I'm being a bit dense, but if this is ensured by the parser, then should this not be just an assert? Or are you saying that this is not ensured by the parser, but an erroneous node here will already have an error report from the parser?

Reply:
In general, the latter. So, if we have an assert, the assertion will throw and omit the properly described error.

For this particular case, I'm not sure if/how this case could be reached, since the grammar requires 'nondet' for this rule. But, if somehow this is possible, the reported error still is much better than an assertion - and there must be a reported error.

Reply:
But, iiuc, if we get here and this is not a `def` them something must be broken with the parser? So assert or exceptions seems justified. Otherwise, e.g., what is something changes upstream that removes generation of an error when this is not a def? Then we would proceed silently and without marking any errer, no?

Reply:
Yep, that sounds good. Thanks.